src/storage.ts

```diff
@@ -0,0 +1,87 @@
+
+import Database from 'better-sqlite3';
+import { createPool as MySQLCreatePool } from 'mysql2';
+import pg from 'pg';
+import Cursor from 'pg-cursor';
+
+const storageTypes = {
+  LEVEL    : 'level',
+  SQLITE   : 'sqlite',
+  MYSQL    : 'mysql',
+  POSTGRES : 'postgres'
+};
+
+function getDBFromURI(u: URL): Dialect {
+  switch(u.protocol.slice(0, -1)) {
+  case storageTypes.SQLITE:
+    return new SqliteDialect({
+      database: async () => new Database(u.host + u.pathname),
+    });
+  case storageTypes.MYSQL:
+    return new MysqlDialect({
+      pool: async () => MySQLCreatePool(u.toString()),
+    });
+  case storageTypes.POSTGRES:
+    return new PostgresDialect({
+      pool   : async () => new pg.Pool({u}),
+      cursor : Cursor,
+    });
+  }
+}
+
+function invalidStorageSchemeMessage(protocol: string): string {
+  let schemes = [];
+  for (const [_, value] of Object.entries(storageTypes)) {
+    schemes.push(value);
+  }
+  return 'Unknown storage protocol ' + protocol.slice(0, 1) + '! Please use one of: ' + schemes.join(', ') + '. For details, see README';
+}
+
+export function getDWNConfig(config: Config): DwnConfig {
+  let dataStore: DataStore;
+  const dataStoreURI = new URL(config.dataStore);
+  switch(dataStoreURI.protocol.slice(0, -1)) {
+  case storageTypes.LEVEL:
+    dataStore = new DataStoreLevel({ blockstoreLocation: dataStoreURI.host + dataStoreURI.pathname + '/DATASTORE' });
+    break;
+  case storageTypes.SQLITE || storageTypes.MYSQL || storageTypes.POSTGRES:
+    dataStore = new DataStoreSql(getDBFromURI(dataStoreURI));
+    break;
+  default:
+    throw invalidStorageSchemeMessage(dataStoreURI.protocol);
+  }
+
+  let eventLog: EventLog;
+  const eventLogURI = new URL(config.eventLog);
+  switch(eventLogURI.protocol.slice(0, -1)) {
+  case storageTypes.LEVEL:
+    eventLog = new EventLogLevel({ location: eventLogURI.host + eventLogURI.pathname + '/EVENTLOG' });
+    break;
+  case storageTypes.SQLITE || storageTypes.MYSQL || storageTypes.POSTGRES:
+    eventLog = new EventLogSql(getDBFromURI(eventLogURI));
+    break;
+  default:
+    throw invalidStorageSchemeMessage(eventLogURI.protocol);
+  }
+
+  let messageStore: MessageStore;
+  const messageStoreURI = new URL(config.messageStore);
+  switch(messageStoreURI.protocol.slice(0, -1)) {
+  case storageTypes.LEVEL:
+    messageStore = new MessageStoreLevel({
+      blockstoreLocation : messageStoreURI.host + messageStoreURI.pathname + '/MESSAGESTORE',
+      indexLocation      : messageStoreURI.host + messageStoreURI.pathname + '/INDEX'
+    });
+    break;
+  case storageTypes.SQLITE || storageTypes.MYSQL || storageTypes.POSTGRES:
+    messageStore = new MessageStoreSql(getDBFromURI(messageStoreURI));
+    break;
+  default:
+    throw invalidStorageSchemeMessage(messageStoreURI.protocol);
+  }
+
+  return { eventLog, dataStore, messageStore };
+}
```

Comment:
So how would you feel on reducing repetition by moving some stuff into helper functions? It's a crude example and we can play around, rename and add enums of course, but you got the idea...

```suggestion
function getLevelStore(storeURI: URL, storeType: string) {
  let levelStore;

  switch (storeType) {
    case 'DATA':
      levelStore = new DataStoreLevel({
        blockstoreLocation: storeURI.host + storeURI.pathname + '/DATASTORE',
      });
      break;

    case 'MESSAGE':
      levelStore = new DataStoreLevel({
        blockstoreLocation: storeURI.host + storeURI.pathname + '/DATASTORE',
      });
      break;

    case 'EVENT':
      levelStore = new EventLogLevel({ location: storeURI.host + storeURI.pathname + '/EVENTLOG' });
      break;

    default:
      throw new Error('Unexpected level store type');
  }

  return levelStore;
}

const getStore = (storeString: string, storeType) => {
  const storeURI = new URL(storeString);
  let store;

  switch (storeURI.protocol.slice(0, -1)) {
    case storageTypes.LEVEL:
      store = getLevelStore(storeURI, storeType);
      break;
    case storageTypes.SQLITE || storageTypes.MYSQL || storageTypes.POSTGRES:
      store = new DataStoreSql(getDBFromURI(storeURI));
      break;
    default:
      throw invalidStorageSchemeMessage(storeURI.protocol);
  }

  return store;
};

export function getDWNConfig(config: Config): DwnConfig {
  let dataStore: DataStore = getStore(config.dataStore, 'DATA');
  let eventLog: EventLog = getStore(config.eventLog, 'EVENT');
  let messageStore: MessageStore = getStore(config.messageStore, 'MESSAGE');

  return { eventLog, dataStore, messageStore };
}
```

Reply:
i would prefer to not have untyped functions at the cost of a little bit of repetition. I don't think several blocks of similar-looking code are really that big a deal. I think this is mostly about my bias in favor of using the typing system wherever possible. I find it makes it easier to catch mistakes quickly.

Reply:
yeah, so we just add types. 
My suggestion was more of a sketch, expecting addition of types/enums of course. 
Let me know if you need help with typing this code.